app/modules/pixbuf.c

```diff
@@ -0,0 +1,555 @@
+  // Allocate memory
+  size_t size = sizeof(pixbuf) + pixbuf_channels_for(type)*leds;
+  pixbuf *buffer = (pixbuf*)lua_newuserdata(L, size);
+
+  // Associate its metatable
+  luaL_getmetatable(L, PIXBUF_METATABLE);
+  lua_setmetatable(L, -2);
+
+  // Save led strip size
+  *(size_t *)&buffer->npix = leds;
+  *(uint8_t *)&buffer->type = type;
+
+  memset(buffer->values, 0, pixbuf_channels_for(type) * leds);
+
+  return buffer;
+}
+
+// Handle a buffer where we can store led values
+int pixbuf_new_lua(lua_State *L) {
+  const int leds = luaL_checkint(L, 1);
+  const int type = luaL_checkint(L, 2);
+
+  luaL_argcheck(L, leds > 0, 1, "should be a positive integer");
+  luaL_argcheck(L, type >= PIXBUF_TYPE_RGB && type < PIXBUF_NTYPES,
+                    2, "must be a valid pixbuf type");
+
+  pixbuf *buffer = pixbuf_new(L, leds, type);
+
+  return 1;
+}
+
+static int pixbuf_concat_lua(lua_State *L) {
+  pixbuf *lhs = pixbuf_from_lua_arg(L, 1);
+  pixbuf *rhs = pixbuf_from_lua_arg(L, 2);
+
+  luaL_argcheck(L, lhs->type == rhs->type, 1,
+                "can only concatenate buffers with same type");
+
+  pixbuf *buffer = pixbuf_new(L, lhs->npix + rhs->npix, lhs->type);
+
+  memcpy(buffer->values, lhs->values, pixbuf_size(lhs));
+  memcpy(buffer->values + pixbuf_size(lhs), rhs->values, pixbuf_size(rhs));
+
+  return 1;
+}
+
+static int pixbuf_dump_lua(lua_State *L) {
+  pixbuf *buffer = pixbuf_from_lua_arg(L, 1);
+  lua_pushlstring(L, buffer->values, pixbuf_size(buffer));
+  return 1;
+}
+
+static int pixbuf_eq_lua(lua_State *L) {
+  bool res;
+
+  pixbuf *lhs = pixbuf_from_lua_arg(L, 1);
+  pixbuf *rhs = pixbuf_from_lua_arg(L, 2);
+
+  if (lhs->npix != lhs->npix) {
+    res = false;
+  } else if (lhs->type != rhs->type) {
+    res = false;
+  } else {
+    res = true;
+    for(size_t i = 0; i < pixbuf_size(lhs); i++) {
+      if(lhs->values[i] != rhs->values[i]) {
+        res = false;
+        break;
+      }
+    }
+  }
+
+  lua_pushboolean(L, res);
+  return 1;
+}
+
+void pixbuf_fade(pixbuf *buffer, int fade, unsigned direction) {
+  // XXX Intensity-bearing types should be handled differently.
+
+  uint8_t *p = &buffer->values[0];
+  int val = 0;
+  int i;
+  for (i = 0; i < pixbuf_size(buffer); i++)
+  {
+    if (direction == PIXBUF_FADE_OUT)
+    {
+      *p++ /= fade;
+    }
+    else
+    {
+      // as fade in can result in value overflow, an int is used to perform the check afterwards
+      val = *p * fade;
+      if (val > 255) val = 255;
+      *p++ = val;
+    }
+  }
+}
+
+static int pixbuf_fade_lua(lua_State *L) {
+  pixbuf *buffer = pixbuf_from_lua_arg(L, 1);
+  const int fade = luaL_checkinteger(L, 2);
+  unsigned direction = luaL_optinteger( L, 3, PIXBUF_FADE_OUT );
+
+  luaL_argcheck(L, fade > 0, 2, "fade value should be a strict positive number");
+
+  pixbuf_fade(buffer, fade, direction);
+
+  return 0;
+}
+
+static int pixbuf_fill_lua(lua_State *L) {
+  pixbuf *buffer = pixbuf_from_lua_arg(L, 1);
```

Comment:
hmm this works but it might me better to have one common function which does all the work which is called with a boolean to know whether to switch colors or not.  That way the cost for maintenance of the code would be less.